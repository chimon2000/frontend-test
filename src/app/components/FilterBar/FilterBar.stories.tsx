import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { boolean, text } from "@storybook/addon-knobs";

import { FilterBar } from "./FilterBar";

const stories = storiesOf("Components", module);

stories.add(
  "Filter Bar",
  withInfo({ inline: true })(() => (
    <FilterBar
      prices={["$", "$$", "$$$"]}
      categories={["French", "Italian", "Japanese"]}
      values={{
        prices: ["$$"],
        categories: ["French"],
        open: false
      }}
      onUpdate={action("Update event")}
    />
  ))
);
