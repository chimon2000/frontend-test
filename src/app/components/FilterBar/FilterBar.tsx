import * as React from "react";
import * as _ from "lodash";

import Dropdown from "../Dropdown";
import Checkbox from "../Checkbox";
import OpenToggleInput from "../OpenToggleInput";

export interface ChangeEventObject {
  name: string;
  value: boolean;
}

export interface FilterValues {
  open: boolean;
  prices: Array<string>;
  categories: Array<string>;
}

export interface FilterBarProps {
  categories: Array<string>;
  prices: Array<string>;
  onUpdate: (name: string, changes: ChangeEventObject) => void;
  values: FilterValues;
}

export const FILTER_DEFAULT: FilterValues = {
  prices: [],
  categories: [],
  open: true
};

export const FilterBar: React.FunctionComponent<FilterBarProps> = ({
  categories,
  prices,
  onUpdate,
  values = FILTER_DEFAULT
}) => {
  return (
    <div className="filter-bar">
      <div className="filter-bar__inline-label">Filter By: </div>
      <div className="filter-bar__inline-input">
        <OpenToggleInput
          className="filter-bar__open-toggle-input"
          id="open-toggle"
          name="is-open"
          checked={values.open}
          onChange={value => onUpdate("open", value)}
        />
      </div>
      <div className="filter-bar__inline-input">
        <Dropdown name="Price" className="filter-bar__price-dropdown-input">
          {prices.map((price: string, idx: number) => (
            <Checkbox
              name={price}
              key={`price-${idx}`}
              id={`price-${idx}`}
              label={price}
              onChange={value => onUpdate("price", value)}
              checked={_.indexOf(values.prices, price) > -1}
            />
          ))}
        </Dropdown>
      </div>
      <div className="filter-bar__inline-input">
        <Dropdown
          name="Categories"
          className="filter-bar__category-dropdown-input"
        >
          {categories.map((category: string, idx: number) => (
            <Checkbox
              name={category}
              key={`category-${idx}`}
              id={`category-${idx}`}
              label={category}
              onChange={value => onUpdate("categories", value)}
              checked={_.indexOf(values.categories, category) > -1}
            />
          ))}
        </Dropdown>
      </div>
    </div>
  );
};
