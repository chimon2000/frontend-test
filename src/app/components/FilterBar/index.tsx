export {
  FilterBar as default,
  ChangeEventObject,
  FilterValues,
  FILTER_DEFAULT
} from "./FilterBar";
