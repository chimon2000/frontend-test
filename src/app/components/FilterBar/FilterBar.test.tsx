import * as React from "react";
import { shallow } from "enzyme";

import { FilterBar } from "./FilterBar";

describe("<FilterBar />", () => {
  it("doesn't blow up", () => {
    const wrapper = shallow(
      <FilterBar
        prices={["$", "$$", "$$$"]}
        categories={["French", "Italian", "Japanese"]}
        values={{
          prices: ["$$"],
          categories: ["French"],
          open: false
        }}
        onUpdate={() => {}}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("prices: sends a name/value update on change", () => {
    let result: { name: string; value: any } = {
      name: null,
      value: {}
    };
    const callback = (name, value) => {
      result = {
        name,
        value
      };
    };

    const wrapper = shallow(
      <FilterBar
        prices={["$", "$$", "$$$"]}
        categories={["French", "Italian", "Japanese"]}
        values={{
          prices: ["$$"],
          categories: ["French"],
          open: false
        }}
        onUpdate={callback}
      />
    );

    const price = wrapper
      .find(".filter-bar__price-dropdown-input")
      .find('[name="$$"]');

    expect(price.prop("checked")).toEqual(true);

    price.simulate("change", { name: "$$", checked: false });

    expect(result.name).toEqual("prices");
    expect(result.value.name).toEqual("$$");
    expect(result.value.checked).toEqual(false);
  });

  it("categories: sends a name/value update on change", () => {
    let result: { name: string; value: any } = {
      name: null,
      value: {}
    };
    const callback = (name, value) => {
      result = {
        name,
        value
      };
    };

    const wrapper = shallow(
      <FilterBar
        prices={["$", "$$", "$$$"]}
        categories={["French", "Italian", "Japanese"]}
        values={{
          prices: ["$$"],
          categories: ["French"],
          open: false
        }}
        onUpdate={callback}
      />
    );

    const category = wrapper
      .find(".filter-bar__category-dropdown-input")
      .find('[name="French"]');

    expect(category.prop("checked")).toEqual(true);

    category.simulate("change", { name: "French", checked: false });

    expect(result.name).toEqual("categories");
    expect(result.value.name).toEqual("French");
    expect(result.value.checked).toEqual(false);
  });

  it("toggles the open boolean on change", () => {
    let result: { name: string; value: any } = {
      name: null,
      value: {}
    };
    const callback = (name, value) => {
      console.log(">>>", name);
      result = {
        name,
        value
      };
    };

    const wrapper = shallow(
      <FilterBar
        prices={["$", "$$", "$$$"]}
        categories={["French", "Italian", "Japanese"]}
        values={{
          prices: ["$$"],
          categories: ["French"],
          open: false
        }}
        onUpdate={callback}
      />
    );
    const open = wrapper
      .find(".filter-bar__open-toggle-input")
      .find('[name="open"]');

    expect(open.prop("checked")).toEqual(false);

    open.simulate("change", { name: "open", checked: true });

    expect(result.name).toEqual("open");
    expect(result.value.name).toEqual("open");
    expect(result.value.checked).toEqual(true);
  });
});
