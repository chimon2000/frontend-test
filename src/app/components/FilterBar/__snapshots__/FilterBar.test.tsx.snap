// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterBar /> doesn't blow up 1`] = `
<div
  className="filter-bar"
>
  <div
    className="filter-bar__inline-label"
  >
    Filter By: 
  </div>
  <div
    className="filter-bar__inline-input"
  >
    <Component
      checked={false}
      className="filter-bar__open-toggle-input"
      id="open-toggle"
      name="open"
      onChange={[Function]}
    />
  </div>
  <div
    className="filter-bar__inline-input"
  >
    <Component
      className="filter-bar__price-dropdown-input"
      name="Price"
    >
      <Component
        checked={false}
        id="price-0"
        key="price-0"
        label="$"
        name="$"
        onChange={[Function]}
      />
      <Component
        checked={true}
        id="price-1"
        key="price-1"
        label="$$"
        name="$$"
        onChange={[Function]}
      />
      <Component
        checked={false}
        id="price-2"
        key="price-2"
        label="$$$"
        name="$$$"
        onChange={[Function]}
      />
    </Component>
  </div>
  <div
    className="filter-bar__inline-input"
  >
    <Component
      className="filter-bar__category-dropdown-input"
      name="Categories"
    >
      <Component
        checked={true}
        id="category-0"
        key="category-0"
        label="French"
        name="French"
        onChange={[Function]}
      />
      <Component
        checked={false}
        id="category-1"
        key="category-1"
        label="Italian"
        name="Italian"
        onChange={[Function]}
      />
      <Component
        checked={false}
        id="category-2"
        key="category-2"
        label="Japanese"
        name="Japanese"
        onChange={[Function]}
      />
    </Component>
  </div>
</div>
`;
